manifests/fact.pp

```diff
@@ -10,6 +10,10 @@
   include ::puppet::defaults
   $facterbasepath = $::puppet::defaults::facterbasepath
 
+  unless is_string($value) or is_array($value) or is_hash($value) {
+    warning("Param \$value of define Puppet::Fact must be Sring or Array or Hash!")
+  }
+
```

Comment:
Can you please add this into the puppet::facts class as well?
You have changed it's functionality as well so this warning is going to be needed there as well.